cmd/dowload_test.go

```diff
@@ -0,0 +1,119 @@
+package cmd
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/cisco-open/grabit/test"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRunDownload(t *testing.T) {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte(`abcdef`))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	port, server := test.HttpHandler(handler)
+	defer server.Close()
+	testfilepath := test.TmpFile(t, fmt.Sprintf(`
+	[[Resource]]
+	Urls = ['http://localhost:%d/test.html']
+	Integrity = 'sha256-vvV+x/U6bUC+tkCngKY5yDvCmsipgW8fxsXG3Nk8RyE='
+`, port))
+	outputDir := test.TmpDir(t)
+	cmd := NewRootCmd()
+	cmd.SetArgs([]string{"-f", testfilepath, "download", "--dir", outputDir})
+	err := cmd.Execute()
+	assert.Nil(t, err)
+}
+
+func TestRunDownloadWithTags(t *testing.T) {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte(`abcdef`))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	port, server := test.HttpHandler(handler)
+	defer server.Close()
+	testfilepath := test.TmpFile(t, fmt.Sprintf(`
+	[[Resource]]
+	Urls = ['http://localhost:%d/test.html']
+	Integrity = 'sha256-vvV+x/U6bUC+tkCngKY5yDvCmsipgW8fxsXG3Nk8RyE='
+	Tags = ['tag']
+
+	[[Resource]]
+	Urls = ['http://localhost:%d/test2.html']
+	Integrity = 'sha256-vvV+x/U6bUC+tkCngKY5yDvCmsipgW8fxsXG3Nk8RyE='
+	Tags = ['tag1', 'tag2']
+`, port, port))
+	outputDir := test.TmpDir(t)
+	cmd := NewRootCmd()
+	cmd.SetArgs([]string{"-f", testfilepath, "download", "--tag", "tag", "--dir", outputDir})
+	err := cmd.Execute()
+	assert.Nil(t, err)
+	files, err := os.ReadDir(outputDir)
+	assert.Nil(t, err)
+	actualFiles := []string{}
+	for _, file := range files {
+		actualFiles = append(actualFiles, file.Name())
+	}
+	assert.ElementsMatch(t, []string{"test.html"}, actualFiles)
+}
+
+func TestRunDownloadWithoutTags(t *testing.T) {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte(`abcdef`))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	port, server := test.HttpHandler(handler)
+	defer server.Close()
+	testfilepath := test.TmpFile(t, fmt.Sprintf(`
+	[[Resource]]
+	Urls = ['http://localhost:%d/test.html']
+	Integrity = 'sha256-vvV+x/U6bUC+tkCngKY5yDvCmsipgW8fxsXG3Nk8RyE='
+	Tags = ['tag']
+
+	[[Resource]]
+	Urls = ['http://localhost:%d/test2.html']
+	Integrity = 'sha256-vvV+x/U6bUC+tkCngKY5yDvCmsipgW8fxsXG3Nk8RyE='
+	Tags = ['tag1', 'tag2']
+`, port, port))
```

Comment:
Is there an easy way to share the code between here and `TestRunDownloadWithTags()`?  It seems largely the same.

Reply:
Good point: I refactored the tests here to add a few utilities that remove most of the duplication (`test.TestHttpHandler` and `test.AssertFileContains`).
For the test data: I'd rather keep the slight duplication to make sure that each test is completely self-contained: each test can be read and understood (preparation, execution, test) without having to look at utility methods to understand what's happening.